### ISSUE TITLE
Support associative arrays when an addXXXX method exists

### ISSUE THREAD

**Issue by simshaun** [nelmio/alice#436]
I'm trying to hard-code some data in a couple fixtures, and it doesn't seem like associative arrays are supported when the object has an `add{XXXXX}` method.

``` yaml
AppBundle\Entity\Bucket:
    bucket_foo:
        name: Foo
        provider: foo
        providerData:
            apiKey: '123'
    bucket_bar:
        name: Bar
        provider: bar
        providerData:
            apiKey: '456'
```

My entity has a convenience method `addProviderData($key, $val)` as well as `setProviderData(array $data)`.

Alice is calling `->addProviderData('123')`, so I think it's dropping the array keys for some reason.

When I remove the `addProviderData` method, Alice calls `->setProviderData(['apiKey' => '123'])` as expected.

**Comment by theofidry** [comment]
After looking at this again, this is the expected behaviour. Keys are not dropped in `setProviderData()`, because it's not alice responsibility to normalize the input, if you inject `['foo' => 'bar']` you should get that array, if you put `['bar']` idem. If you want a guarantee to have a clean array you should normalize it.

In 3.x, this behaviour won't change, you will have easier time to customize it if you want to [custom accessor](https://github.com/nelmio/alice/blob/master/doc/advanced-guide.md#custom-accessor), but the default behaviour will remain unchanged.

**Comment by simshaun** [comment]
It's been a while since I've dealt with this, but this problem/bug didn't have anything to do with `setProviderData` other than Alice was choosing to call `addProviderData` over it.

The problem with Alice calling `addProviderData` is that it wasn't passing the key *(apiKey)* to the method. It was only passing the value *(456)* to it. I needed both the key and the value.

The current Alice behavior is fine if your *addXXXX* methods simply add an element to a list, but it breaks down if your *addXXXX* methods are adding a key/value on a map.

There needs to be some way to tell Alice to pass either 1) both key & value, or 2 ) just value.

**Comment by theofidry** [comment]
I understand the problem. In 2.x I don't think this can be changed, this would be a BC break although I don't expect many people to encounter it.

In 3.x, Alice is relying on Symfony PropertyAccess for that as AFAIK it doesn't support this case of `add($key, $value)` out of the box like that either. It is however, far easier to support it. But as they are some performance impact (it requires to analyse the adder signature via reflection), I wouldn't do it out of the box either, rather let the user use a point of extension to deal with this issue (and the perf impact is likely to be way smaller that way as well, as you can go for a more in-house solution rather than a generic one).

**Comment by simshaun** [comment]
When there's a map in the YAML, e.g.

    providerData:
        apiKey: '123'
        version: '2.x'

Is there any way to have Alice just call `setProviderData` instead of `addProviderData`? That would mitigate the problem.

If it's a list, then call `addXXXX`:

    people:
        - 'John'
        - 'Jane'

**Comment by theofidry** [comment]
>Is there any way to have Alice just call setProviderData instead of addProviderData?

Yes, in 2.x by re-ordering the populator methods, but that's rather painful. In 3.x it's easy to extend it via the link I gave you.

I do understand it may not be the best thing in the world there, but the alternative is, besides technically difficult and complex, not something everyone wants either. I prefer to keep this behaviour consistent between 2.x and 3.x and just make it easier for people to change something at this level rather (which is the link I gave you).

As for the associative array/list problem, it is hard, PHP does not have proper structures for it so a way would be to do a check if the keys of the array and `array_keys(array_values($x))` would be identical, doable but not trivial.
